File: src/csv_exporter.py
"""
CSV exporter for economic events
"""

import pandas as pd
import logging
from datetime import datetime
from typing import List, Dict, Optional
import os

logger = logging.getLogger(__name__)


class CSVExporter:
    """Exports economic events to CSV format"""
    
    def __init__(self, output_path: str = "output/economic_events.csv"):
        self.output_path = output_path
        self._ensure_output_directory()
    
    def _ensure_output_directory(self):
        """Ensure the output directory exists"""
        output_dir = os.path.dirname(self.output_path)
        if output_dir and not os.path.exists(output_dir):
            os.makedirs(output_dir)
            logger.info(f"Created output directory: {output_dir}")
    
    def export_events(self, events: List[Dict], mode: str = 'w') -> bool:
        """
        Export economic events to CSV file
        
        Args:
            events: List of event dictionaries
            mode: Write mode ('w' for overwrite, 'a' for append)
            
        Returns:
            True if successful, False otherwise
        """
        if not events:
            logger.warning("No events to export")
            return False
        
        try:
            # Convert to DataFrame
            df = pd.DataFrame(events)
            
            # Ensure required columns exist
            required_columns = [
                'DateTime', 'Event', 'Country', 'Impact', 
                'Currency', 'Actual', 'Forecast', 'Previous', 'AffectedPairs'
            ]
            
            # Add missing columns with default values
            for col in required_columns:
                if col not in df.columns:
                    df[col] = 'N/A'
                    logger.warning(f"Missing column {col}, filled with 'N/A'")
            
            # Reorder columns to match expected format
            df = df[required_columns]
            
            # Convert DateTime to string format for CSV and validate
            if 'DateTime' in df.columns:
                df['DateTime'] = pd.to_datetime(df['DateTime']).dt.strftime('%Y-%m-%d %H:%M:%S')
            
            # Validate data quality
            self._validate_data_quality(df)
            
            # Handle append mode
            if mode == 'a':
                if os.path.exists(self.output_path):
                    # File exists, read it and append
                    existing_df = pd.read_csv(self.output_path)
                    df = pd.concat([existing_df, df], ignore_index=True)
                    # Remove duplicates based on DateTime, Event, and Currency
                    df = self._remove_duplicates(df)
                    # Write the complete updated file
                    df.to_csv(self.output_path, index=False, mode='w')
                else:
                    # File doesn't exist yet, create it with headers
                    df.to_csv(self.output_path, index=False, mode='w')
            else:
                # Write mode (overwrite)
                df.to_csv(self.output_path, index=False, mode=mode)
            
            logger.info(f"Exported {len(df)} events to {self.output_path}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to export events: {e}")
            return False
    
    def _validate_data_quality(self, df: pd.DataFrame) -> None:
        """
        Validate data quality and log statistics
        
        Args:
            df: DataFrame to validate
        """
        if df.empty:
            logger.warning("No data to validate")
            return
        
        try:
            total_events = len(df)
            
            # Check for missing critical data
            missing_event_names = df['Event'].isna().sum()
            missing_currencies = df['Currency'].isna().sum()
            missing_dates = df['DateTime'].isna().sum()
            
            # Check data quality metrics
            events_with_actual = (df['Actual'] != 'N/A').sum()
            events_with_forecast = (df['Forecast'] != 'N/A').sum()
            events_with_previous = (df['Previous'] != 'N/A').sum()
            
            # Count by impact level
            impact_counts = df['Impact'].value_counts().to_dict()
            
            # Count by currency
            currency_counts = df['Currency'].value_counts().head(10).to_dict()
            
            logger.info("=" * 50)
            logger.info("DATA QUALITY VALIDATION REPORT")
            logger.info("=" * 50)
            logger.info(f"[STATS] Total events: {total_events}")
            logger.info(f"[CHECK] Missing Event names: {missing_event_names}")
            logger.info(f"[CURRENCY] Missing Currencies: {missing_currencies}")
            logger.info(f"[DATE] Missing Dates: {missing_dates}")
            logger.info(f"[ACTUAL] Events with Actual values: {events_with_actual} ({events_with_actual/total_events*100:.1f}%)")
            logger.info(f"[FORECAST] Events with Forecast values: {events_with_forecast} ({events_with_forecast/total_events*100:.1f}%)")
            logger.info(f"[PREVIOUS] Events with Previous values: {events_with_previous} ({events_with_previous/total_events*100:.1f}%)")
            
            logger.info("Impact Distribution:")
            for impact, count in impact_counts.items():
                logger.info(f"  {impact}: {count} ({count/total_events*100:.1f}%)")
            
            logger.info("Top Currencies:")
            for currency, count in currency_counts.items():
                logger.info(f"  {currency}: {count}")
            
            logger.info("=" * 50)
            
        except Exception as e:
            logger.warning(f"Error during data validation: {e}")
    
    def _remove_duplicates(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        Remove duplicate events based on DateTime, Event, and Currency
        
        Args:
            df: DataFrame to deduplicate
            
        Returns:
            Deduplicated DataFrame
        """
        if df.empty:
            return df
        
        # Sort by DateTime to keep the most recent version
        df = df.sort_values('DateTime', ascending=False)
        
        # Remove duplicates keeping the first occurrence (most recent)
        df = df.drop_duplicates(
            subset=['DateTime', 'Event', 'Currency'], 
            keep='first'
        )
        
        # Sort back by DateTime ascending
        df = df.sort_values('DateTime', ascending=True)
        
        return df
    
    def append_events(self, events: List[Dict]) -> bool:
        """
        Append events to existing CSV file
        
        Args:
            events: List of event dictionaries to append
            
        Returns:
            True if successful, False otherwise
        """
        return self.export_events(events, mode='a')
    
    def get_existing_events(self) -> Optional[pd.DataFrame]:
        """
        Load existing events from CSV file
        
        Returns:
            DataFrame of existing events or None if file doesn't exist
        """
        try:
            if os.path.exists(self.output_path):
                df = pd.read_csv(self.output_path)
                # Convert DateTime back to datetime
                if 'DateTime' in df.columns:
                    df['DateTime'] = pd.to_datetime(df['DateTime'])
                return df
            else:
                logger.info(f"Output file {self.output_path} does not exist")
                return None
        except Exception as e:
            logger.error(f"Failed to read existing events: {e}")
            return None
    
    def backup_existing_file(self) -> bool:
        """
        Create a backup of the existing CSV file
        
        Returns:
            True if successful, False otherwise
        """
        if not os.path.exists(self.output_path):
            logger.info("No existing file to backup")
            return True
        
        try:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            backup_path = f"{self.output_path}.backup_{timestamp}"
            
            import shutil
            shutil.copy2(self.output_path, backup_path)
            logger.info(f"Created backup: {backup_path}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to create backup: {e}")
            return False
    
    def append_with_deduplication(self, new_events: List[Dict]) -> bool:
        """
        Append new events to CSV with deduplication based on DateTime + Event
        
        Args:
            new_events: List of new event dictionaries to append
            
        Returns:
            True if successful, False otherwise
        """
        if not new_events:
            logger.warning("No new events to append")
            return False
        
        try:
            # Read existing events if file exists
            existing_events = []
            if os.path.exists(self.output_path):
                try:
                    existing_df = pd.read_csv(self.output_path)
                    existing_events = existing_df.to_dict('records')
                    logger.info(f"Loaded {len(existing_events)} existing events from CSV")
                except Exception as e:
                    logger.warning(f"Could not read existing CSV file: {e}")
                    existing_events = []
            
            # Create a dictionary with key = (DateTime, Event) for deduplication
            events_dict = {}
            
            # Add existing events to dictionary
            for event in existing_events:
                key = (event.get('DateTime', ''), event.get('Event', ''))
                if key[0] and key[1]:  # Only add if both DateTime and Event exist
                    events_dict[key] = event
            
            # Add/update with new events
            updated_count = 0
            added_count = 0
            holiday_events = [e for e in new_events if e.get('Impact') == 'Holiday']
            logger.info(f"Processing {len(holiday_events)} Holiday events in new events")
            
            for event in new_events:
                key = (event.get('DateTime', ''), event.get('Event', ''))
                if key[0] and key[1]:  # Only process if both DateTime and Event exist
                    if key in events_dict:
                        # Update existing event
                        events_dict[key] = event
                        updated_count += 1
                        if event.get('Impact') == 'Holiday':
                            logger.debug(f"Updated Holiday event: {key[1]} on {key[0]}")
                    else:
                        # Add new event
                        events_dict[key] = event
                        added_count += 1
                        if event.get('Impact') == 'Holiday':
                            logger.debug(f"Added Holiday event: {key[1]} on {key[0]}")
                        logger.debug(f"Added new event: {key[1]} on {key[0]}")
            
            # Convert back to list and sort by DateTime (newest first)
            all_events = list(events_dict.values())
            
            # Count holidays in final events
            final_holidays = [e for e in all_events if e.get('Impact') == 'Holiday']
            logger.info(f"Final events after deduplication: {len(all_events)} total, {len(final_holidays)} holidays")
            
            # Sort by DateTime (newest first)
            try:
                all_events.sort(key=lambda x: pd.to_datetime(x.get('DateTime', '1900-01-01')), reverse=True)
            except Exception as e:
                logger.warning(f"Could not sort events by DateTime: {e}")
            
            # Export the deduplicated and sorted events
            success = self.export_events(all_events, mode='w')  # Use 'w' to overwrite with deduplicated data
            
            if success:
                logger.info(f"Deduplication complete: {added_count} new events added, {updated_count} events updated")
                logger.info(f"Total events in CSV: {len(all_events)}")
            
            return success
            
        except Exception as e:
            logger.error(f"Failed to append events with deduplication: {e}")
            return False

    def get_file_info(self) -> Dict[str, str]:
        """
        Get information about the output file
        
        Returns:
            Dictionary with file information
        """
        info = {
            'path': self.output_path,
            'exists': str(os.path.exists(self.output_path)),
            'size': '0 bytes'
        }
        
        try:
            if os.path.exists(self.output_path):
                size_bytes = os.path.getsize(self.output_path)
                info['size'] = f"{size_bytes} bytes"
                info['last_modified'] = str(datetime.fromtimestamp(os.path.getmtime(self.output_path)))
        except Exception as e:
            logger.warning(f"Could not get file info: {e}")
        
        return info

File: JTrading.NewsManager.CSharp/src/Services/CsvExporter.cs
using System.Globalization;
using CsvHelper;
using CsvHelper.Configuration;
using Microsoft.Extensions.Logging;
using JTrading.NewsManager.Domain.Models;

namespace JTrading.NewsManager.Services;

public class CsvExporter
{
    private readonly string _outputPath;
    private readonly ILogger<CsvExporter>? _logger;

    public CsvExporter(string outputPath, ILogger<CsvExporter>? logger = null)
    {
        _outputPath = outputPath;
        _logger = logger;
        EnsureOutputDirectory();
    }

    private void EnsureOutputDirectory()
    {
        var outputDir = Path.GetDirectoryName(_outputPath);
        if (!string.IsNullOrEmpty(outputDir) && !Directory.Exists(outputDir))
        {
            Directory.CreateDirectory(outputDir);
            _logger?.LogInformation("Created output directory: {Directory}", outputDir);
        }
    }

    public bool ExportEvents(List<EconomicEvent> events, string mode = "w")
    {
        if (events == null || !events.Any())
        {
            _logger?.LogWarning("No events to export");
            return false;
        }

        try
        {
            // Ensure required fields exist
            var eventsWithDefaults = events.Select(evt => new EconomicEvent(evt)
            {
                Actual = string.IsNullOrEmpty(evt.Actual) ? "N/A" : evt.Actual,
                Forecast = string.IsNullOrEmpty(evt.Forecast) ? "N/A" : evt.Forecast,
                Previous = string.IsNullOrEmpty(evt.Previous) ? "N/A" : evt.Previous,
                AffectedPairs = string.IsNullOrEmpty(evt.AffectedPairs) ? "N/A" : evt.AffectedPairs
            }).ToList();

            // Handle append mode
            if (mode == "a" && File.Exists(_outputPath))
            {
                // Read existing events
                var existingEvents = GetExistingEvents();
                if (existingEvents != null && existingEvents.Any())
                {
                    eventsWithDefaults.AddRange(existingEvents);
                    eventsWithDefaults = RemoveDuplicates(eventsWithDefaults);
                    // Write complete updated file
                    WriteEventsToFile(eventsWithDefaults);
                }
                else
                {
                    WriteEventsToFile(eventsWithDefaults);
                }
            }
            else
            {
                // Write mode (overwrite)
                WriteEventsToFile(eventsWithDefaults);
            }

            _logger?.LogInformation("Exported {Count} events to {Path}", eventsWithDefaults.Count, _outputPath);
            return true;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to export events");
            return false;
        }
    }

    private void WriteEventsToFile(List<EconomicEvent> events)
    {
        var config = new CsvConfiguration(CultureInfo.InvariantCulture)
        {
            HasHeaderRecord = true
        };

        using var writer = new StreamWriter(_outputPath, false, System.Text.Encoding.UTF8);
        using var csv = new CsvWriter(writer, config);
        csv.Context.RegisterClassMap<EconomicEventMap>();
        csv.WriteRecords(events);
    }

    public bool AppendEvents(List<EconomicEvent> events)
    {
        return ExportEvents(events, mode: "a");
    }

    public List<EconomicEvent>? GetExistingEvents()
    {
        try
        {
            if (!File.Exists(_outputPath))
            {
                _logger?.LogInformation("Output file {Path} does not exist", _outputPath);
                return null;
            }

            var config = new CsvConfiguration(CultureInfo.InvariantCulture)
            {
                HasHeaderRecord = true
            };

            using var reader = new StreamReader(_outputPath, System.Text.Encoding.UTF8);
            using var csv = new CsvReader(reader, config);
            csv.Context.RegisterClassMap<EconomicEventMap>();

            var events = new List<EconomicEvent>();
            csv.Read();
            csv.ReadHeader();

            while (csv.Read())
            {
                try
                {
                    var evt = csv.GetRecord<EconomicEvent>();
                    if (evt != null)
                    {
                        events.Add(evt);
                    }
                }
                catch (Exception ex)
                {
                    _logger?.LogWarning(ex, "Error reading CSV record at line {Line}", csv.Parser.Row);
                }
            }

            return events;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to read existing events");
            return null;
        }
    }

    private List<EconomicEvent> RemoveDuplicates(List<EconomicEvent> events)
    {
        if (events == null || !events.Any())
        {
            return events ?? new List<EconomicEvent>();
        }

        // Sort by DateTime descending to keep most recent version
        var sorted = events.OrderByDescending(e => e.DateTime).ToList();

        // Remove duplicates based on DateTime, Event, and Currency
        var uniqueEvents = sorted
            .GroupBy(e => new { e.DateTime, e.Event, e.Currency })
            .Select(g => g.First())
            .OrderBy(e => e.DateTime)
            .ToList();

        return uniqueEvents;
    }

    public bool AppendWithDeduplication(List<EconomicEvent> newEvents)
    {
        if (newEvents == null || !newEvents.Any())
        {
            _logger?.LogWarning("No new events to append");
            return false;
        }

        try
        {
            // Read existing events if file exists
            var existingEvents = new List<EconomicEvent>();
            if (File.Exists(_outputPath))
            {
                var loaded = GetExistingEvents();
                if (loaded != null)
                {
                    existingEvents = loaded;
                    _logger?.LogInformation("Loaded {Count} existing events from CSV", existingEvents.Count);
                }
            }

            // Create dictionary for deduplication (key = DateTime + Event + Currency)
            var eventsDict = new Dictionary<string, EconomicEvent>();

            // Add existing events
            foreach (var evt in existingEvents)
            {
                var key = $"{evt.DateTime:yyyy-MM-dd HH:mm:ss}|{evt.Event}|{evt.Currency}";
                if (!string.IsNullOrEmpty(evt.DateTime.ToString()) && !string.IsNullOrEmpty(evt.Event))
                {
                    eventsDict[key] = evt;
                }
            }

            // Add/update with new events
            int updatedCount = 0;
            int addedCount = 0;
            var holidayEvents = newEvents.Where(e => e.Impact == "Holiday").ToList();
            _logger?.LogInformation("Processing {Count} Holiday events in new events", holidayEvents.Count);

            foreach (var evt in newEvents)
            {
                var key = $"{evt.DateTime:yyyy-MM-dd HH:mm:ss}|{evt.Event}|{evt.Currency}";
                if (!string.IsNullOrEmpty(evt.DateTime.ToString()) && !string.IsNullOrEmpty(evt.Event))
                {
                    if (eventsDict.ContainsKey(key))
                    {
                        eventsDict[key] = evt;
                        updatedCount++;
                        if (evt.Impact == "Holiday")
                        {
                            _logger?.LogDebug("Updated Holiday event: {Event} on {DateTime}", evt.Event, evt.DateTime);
                        }
                    }
                    else
                    {
                        eventsDict[key] = evt;
                        addedCount++;
                        if (evt.Impact == "Holiday")
                        {
                            _logger?.LogDebug("Added Holiday event: {Event} on {DateTime}", evt.Event, evt.DateTime);
                        }
                    }
                }
            }

            // Convert back to list and sort by DateTime (newest first for validation)
            var allEvents = eventsDict.Values.OrderByDescending(e => e.DateTime).ToList();

            var finalHolidays = allEvents.Where(e => e.Impact == "Holiday").ToList();
            _logger?.LogInformation(
                "Final events after deduplication: {Total} total, {Holidays} holidays",
                allEvents.Count, finalHolidays.Count);

            // Sort by DateTime ascending for export
            allEvents = allEvents.OrderBy(e => e.DateTime).ToList();

            // Export the deduplicated events
            var success = ExportEvents(allEvents, mode: "w");

            if (success)
            {
                _logger?.LogInformation(
                    "Deduplication complete: {Added} new events added, {Updated} events updated. Total events in CSV: {Total}",
                    addedCount, updatedCount, allEvents.Count);
            }

            return success;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "Failed to append events with deduplication");
            return false;
        }
    }

    public void ValidateDataQuality(List<EconomicEvent> events)
    {
        if (events == null || !events.Any())
        {
            _logger?.LogWarning("No data to validate");
            return;
        }

        try
        {
            var totalEvents = events.Count;

            // Check for missing critical data
            var missingEventNames = events.Count(e => string.IsNullOrEmpty(e.Event));
            var missingCurrencies = events.Count(e => string.IsNullOrEmpty(e.Currency));
            var missingDates = events.Count(e => e.DateTime == default);

            // Check data quality metrics
            var eventsWithActual = events.Count(e => e.Actual != "N/A" && !string.IsNullOrEmpty(e.Actual));
            var eventsWithForecast = events.Count(e => e.Forecast != "N/A" && !string.IsNullOrEmpty(e.Forecast));
            var eventsWithPrevious = events.Count(e => e.Previous != "N/A" && !string.IsNullOrEmpty(e.Previous));

            // Count by impact level
            var impactCounts = events.GroupBy(e => e.Impact)
                .ToDictionary(g => g.Key, g => g.Count());

            // Count by currency
            var currencyCounts = events.GroupBy(e => e.Currency)
                .OrderByDescending(g => g.Count())
                .Take(10)
                .ToDictionary(g => g.Key, g => g.Count());

            _logger?.LogInformation("=" + new string('=', 50));
            _logger?.LogInformation("DATA QUALITY VALIDATION REPORT");
            _logger?.LogInformation("=" + new string('=', 50));
            _logger?.LogInformation("[STATS] Total events: {Total}", totalEvents);
            _logger?.LogInformation("[CHECK] Missing Event names: {Count}", missingEventNames);
            _logger?.LogInformation("[CURRENCY] Missing Currencies: {Count}", missingCurrencies);
            _logger?.LogInformation("[DATE] Missing Dates: {Count}", missingDates);
            _logger?.LogInformation(
                "[ACTUAL] Events with Actual values: {Count} ({Percentage:F1}%)",
                eventsWithActual, eventsWithActual * 100.0 / totalEvents);
            _logger?.LogInformation(
                "[FORECAST] Events with Forecast values: {Count} ({Percentage:F1}%)",
                eventsWithForecast, eventsWithForecast * 100.0 / totalEvents);
            _logger?.LogInformation(
                "[PREVIOUS] Events with Previous values: {Count} ({Percentage:F1}%)",
                eventsWithPrevious, eventsWithPrevious * 100.0 / totalEvents);

            _logger?.LogInformation("Impact Distribution:");
            foreach (var kvp in impactCounts)
            {
                _logger?.LogInformation("  {Impact}: {Count} ({Percentage:F1}%)",
                    kvp.Key, kvp.Value, kvp.Value * 100.0 / totalEvents);
            }

            _logger?.LogInformation("Top Currencies:");
            foreach (var kvp in currencyCounts)
            {
                _logger?.LogInformation("  {Currency}: {Count}", kvp.Key, kvp.Value);
            }

            _logger?.LogInformation("=" + new string('=', 50));
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(ex, "Error during data validation");
        }
    }

    public Dictionary<string, string> GetFileInfo()
    {
        var info = new Dictionary<string, string>
        {
            ["path"] = _outputPath,
            ["exists"] = File.Exists(_outputPath).ToString(),
            ["size"] = "0 bytes"
        };

        try
        {
            if (File.Exists(_outputPath))
            {
                var fileInfo = new FileInfo(_outputPath);
                info["size"] = $"{fileInfo.Length} bytes";
                info["last_modified"] = fileInfo.LastWriteTime.ToString();
            }
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(ex, "Could not get file info");
        }

        return info;
    }
}

